connector_ldap: Implement connection pooling for LDAP connections

fnordahl opened the pull request:
The initial implementation of the LDAP connector initiates and tears down a connection for each session. Although the creation and tear-down of connections are light weight operations for most LDAP server implementations, it is not hard to imagine that this at some point will become a bottle neck and scaling issue.

This PR addresses this issue by implementing connection pooling for the LDAP connections. The pooled connections are checked for liveliness and kept alive. The number of connections to keep as well as the interval for liveliness checks are configurable.

Fixes #309

iyee commented on connector/connector_ldap.go:
```diff
@@ -154,11 +157,29 @@ func (cfg *LDAPConnectorConfig) Connector(ns url.URL, lf oidc.LoginFunc, tpls *t
 	}
 
+	var initConn, maxConn int
+	if cfg.InitConn != 0 {
```
I guess checking whether the config param `> 0` is better.

fnordahl replied:
Yes, thx!

ericchiang commented on Documentation/connectors-configuration.md:
```diff
@@ -167,6 +167,10 @@ In addition to `id` and `type`, the `ldap` connector takes the following additio
 
 * skipCertVerification: a `boolean`. Skip server certificate chain verification.
 
+* initConn: a `integer`. Number of LDAP Connections to establish on startup. Default: `5`
```
I don't think this is necessary. Let's just let the pool fill as various connections are made.

fnordahl replied:
It probably will be good enough, yes. LDAP servers are generally really good at connection initialization and teardown.

ericchiang commented on connector/connector_ldap.go:
```diff
@@ -217,11 +238,74 @@ func (c *LDAPConnector) TrustedEmailProvider() bool {
+		return conn, nil
+	}
+	// the pool is empty
+	conn, err := p.ldapConnect()
```
Because you're holding the mutex here you won't be able to connect more than one connection at a time. Need to unlock before this point.

fnordahl replied:
Yes of course, thanks. One path could be to allocate and store a new ldap.Conn in the map while holding the mutex, release the mutex and then initialize/dial the connection. If that is not allowed by the library I will have to reverse it. Create+Dial the connection and then lock+store+unlock and return. Will fix it.

ericchiang replied:
Maybe this?

```go
func (p *LDAPPool) connFromPool() *ldap.Conn {
    p.mu.Lock()
    defer p.mu.Unlock()

    for conn := range p.conns { // you can range over a nil map
        delete(p.conns, conn)
        return conn
    }
    return nil
}

func (p *LDAPPool) Get() (*ldap.Conn, error) {
    if conn := p.connFromPool(); conn != nil {
        return conn, nil
    }
    return p.ldapConnect()
}
```

fnordahl replied:
Nice. Elegant solution!

ericchiang replied:
Go's net/http.Transport has some pretty neat tricks for [connection caching](https://github.com/golang/go/blob/7c22af830a72adf30b18ee3d1744aab8e3c009a8/src/net/http/transport.go#L668) if you're looking for inspiration. Though this definitely doesn't need to be a sophisticated as that implementation.

ericchiang commented on connector/connector_ldap.go:
```diff
@@ -217,11 +238,74 @@ func (c *LDAPConnector) TrustedEmailProvider() bool {
+		c.Close()
+	} else {
+		// drop to anonymous bind
+		err := c.Bind("", "")
```
Like the comment above, this bind should be performed outside of the mutex lock. The mutex should be used to guard the map and the counter, don't make remote calls while holding it.

fnordahl replied:
Duly noted. Will fix it.

Can we go back to process based multi-processing, please? Forks are almost as cheap as threads these days! ;-) (Do not take this comment too seriously, although forks are cheap and in some circumstances adds to security.)

ericchiang commented on connector/connector_ldap.go:
```diff
@@ -217,11 +238,74 @@ func (c *LDAPConnector) TrustedEmailProvider() bool {
 
+type LDAPPool struct {
+	m          sync.Mutex
+	conns      map[*ldap.Conn]*ldap.Conn
```
Does this work instead?

```
conns map[*ldap.Conn]struct{} // A bag of connections
```

fnordahl replied:
I did consider to change this given that I only use the key of the map throughout the code. The value stored could be nil, so using a pointer type that uses the least amount of memory is probably the thing to look for.

Is your proposal the norm in Go for such usage of maps?

ericchiang replied:
The empty struct takes [zero memory](https://play.golang.org/p/BR2urbScRB). But I wouldn't worry about memory constraints since you'll have at most a 100 connections or so.

I suggested this because `map[*ldap.Conn]*ldap.Conn` looks odd. This just seems more explicit. Wouldn't say it's the norm but I have seen maps combined with the empty struct for things like this.

fnordahl replied:
From small streams, mighty rivers can flow.

Your suggestion, `conns map[*ldap.Conn]struct{}`, looks neat and you can't do much better than zero. I'll use it.

ericchiang commented on connector/connector_ldap.go:
```diff
@@ -154,11 +157,29 @@ func (cfg *LDAPConnectorConfig) Connector(ns url.URL, lf oidc.LoginFunc, tpls *t
+	} else {
+		initConn = 5
+	}
+	if cfg.MaxConn != 0 {
```
Can you clean up these initializations to: 

```
const defaultMaxConns = 30

...

if cfg.MaxConn <=0 {
    cfg.MaxConn = defaultMaxConn
}

...

ldapPool := &LDAPPool{
    ...
    maxConn: cfg.MaxConn,
    ...
}
```

fnordahl replied:
I have deliberately created a temporary variable here and other places in the constructor to avoid modifying cfg.

My reasoning behind this is:

1. Do not to modify data structures created by parts of the code that I do not control directly.
2. If we at one point want to implement online reconfiguration of dex and all its connectors we probably do need to know the previous state of cfg to have something to compare with values read from disk to be able to determine whether a update is necessary for each value.

If you still do not agree with it I will change it, but at least you know the thought behind it.

ericchiang replied:
Sure sounds reasonable. I would still like to see constants for default values.

fnordahl replied:
Yes, there will be constants. I'll post a follow-up PR to turn the other defaults into constants as well.

fnordahl commented:
This probably should have some sort of keep-alive, reaper after N secs or periodic liveliness-check of the connections in the pool.

Maybe a check before return from Get() if the connection comes from the pool would do?

It would be bad to let random end users suffer from dead-after-idle-too-long connections left over from a previous surge in traffic.

ericchiang commented:
Is there a ping equivalent for LDAP connections?

fnordahl commented:
Yes, that + querying the root DSE is considered a "ping". I had it in my original Healthy function. Have it stashed somewhere.

Do we have an existing framework for firing functions periodically? When I think about it there probably should be a maxIdleConn parameter too. No need to keep 30 connections open when there is no pressure.

The connection is already anonymously bound when put in the pool, so I could implement a ping function that queries the root DSE and use it before returning the connection to the caller in Get.

fnordahl commented:
A follow-up question to this would be, do you think we need this for other stuff as well? There is ample opportunity to write a generic engine that can pool and keep-alive/ping anything, given that there already is or probably will be need for it.

ericchiang commented:
I think [`Sync()`](https://github.com/coreos/dex/blob/aff47d2cb2d43b8459fae8f4459bef9c85b99db9/connector/interface.go#L31-L34) is what you're looking for.

fnordahl commented:
Updated version:
- [x] Changed configuration statement from maxConn to maxIdleConn to better reflect what the pool actually does. (There is no cap on how many connections can be made, only how many idle to keep).
- [x] Implemented functions for pinging LDAP connections.
- [x] Implemented periodic liveness check of LDAP connections with configurable timer.

ericchiang commented on connector/connector_ldap.go:
```diff
@@ -217,11 +250,108 @@ func (c *LDAPConnector) TrustedEmailProvider() bool {
+	TlsConfig      *tls.Config
+}
+
+func (p *LDAPPool) Get(params ...*ldap.Conn) (*ldap.Conn, error) {
```
Can you separate getting a random connection (Get()) and getting a specific connection (Get(params)) into different methods? They have very different uses and it's odd to overload this method.

ericchiang commented on connector/connector_ldap.go:
```diff
@@ -217,11 +250,108 @@ func (c *LDAPConnector) TrustedEmailProvider() bool {
+func (p *LDAPPool) Get(params ...*ldap.Conn) (*ldap.Conn, error) {
+	var conn *ldap.Conn
+
+	p.m.Lock()
```
What's being guarded by the lock and what's intentionally outside the lock isn't obvious from this flow, a lot of conditions, and manual unlocks. Splitting out the functionality of Get() and Get(conn) should help.

ericchiang commented on connector/connector_ldap.go:
```diff
@@ -217,11 +250,108 @@ func (c *LDAPConnector) TrustedEmailProvider() bool {
+	// Iterate over snapshot, Get and ping connections.
+	// Put live connections back into pool, Close dead ones.
+	for _, el := range conns {
+		conn, err := p.Get(el)
```
err is being ignored here (I think intentionally). Yeah, this is a bad use for Get.

Maybe just use

```go
// getConn attempts to remove the provided connection from the pool. If getConn returns false
// another routine is using the connection and the caller should discard the pointer.
func (p *LDAPPool) getConn(conn *ldap.Conn) bool {
    p.mu.Lock()
    defer p.mu.Lock()

    _, ok := p.conns[conn]
    delete(p.conns, conn)
    return ok
}
```

fnordahl replied:
Nice suggestion. My original thought was that having two functions that needed to do the locking right would lead to double the opportunity to at some point do it wrong. But I guess that the complexity of the dual use in the one function does just this by itself. I'll split them up and post a new patch set.

Thanks!

bobbyrullo commented on connector/connector_ldap.go:
```diff
@@ -210,18 +231,131 @@ func (c *LDAPConnector) Register(mux *http.ServeMux, errorURL url.URL) {
+			time.Sleep(c.idp.ldapPool.PoolCheckTimer)
+		}
+	}()
 	return make(chan struct{})
```
The return channel should be a stop chan, so that when you poke it, your process stops. (see connector_oidc.go)

fnordahl replied:
Thx, I'll have a look at connector_oidc and make a new attempt at making it work.

bobbyrullo commented on connector/connector_ldap.go:
```diff
@@ -154,11 +159,27 @@ func (cfg *LDAPConnectorConfig) Connector(ns url.URL, lf oidc.LoginFunc, tpls *t
+		ServerPort:     cfg.ServerPort,
+		UseTLS:         cfg.UseTLS,
+		UseSSL:         cfg.UseSSL,
+		TlsConfig:      tlsConfig,
```
nit: TLSConfig

fnordahl replied:
Ack

bobbyrullo commented on connector/connector_ldap.go:
```diff
@@ -210,18 +231,131 @@ func (c *LDAPConnector) Register(mux *http.ServeMux, errorURL url.URL) {
+
+// Ping attempts to iterate over all the connections in the pool and check wheter they are
+// alive or not. Live connections are put back into the pool, dead ones are discarded.
+func (p *LDAPPool) Ping() (int, int) {
```
With a name like `Ping`, I don't expect to see mutating behavior. Could you rename to something like "PruneConnections" or "CheckConnections"

fnordahl replied:
I see your point. I will fix it.

bobbyrullo commented on connector/connector_ldap.go:
```diff
@@ -210,18 +231,131 @@ func (c *LDAPConnector) Register(mux *http.ServeMux, errorURL url.URL) {
+	TlsConfig      *tls.Config
+}
+
+// Get returns a random connection from the pool. A new connection is returned if there are
```
It's not exactly random is it, just picks the first one off the top right?

bobbyrullo commented on connector/connector_ldap.go:
```diff
@@ -210,18 +231,131 @@ func (c *LDAPConnector) Register(mux *http.ServeMux, errorURL url.URL) {
+}
+
+// Put makes a connection ready for re-use and puts it back into the pool. If the connection
+// cannot be reused it is discarded. If there allready are MaxIdleConn connections in the pool
```
spelling: already

fnordahl replied:
Ack

bobbyrullo commented on connector/connector_ldap.go:
```diff
@@ -210,18 +231,131 @@ func (c *LDAPConnector) Register(mux *http.ServeMux, errorURL url.URL) {
+
+// getConn attempts to remove the provided connection from the pool. If getConn returns false
+// another routine is using the connection and the caller should discard the pointer.
+func (p *LDAPPool) getConn(conn *ldap.Conn) bool {
```
similar issue here: you don't expect side-effects with `getXXX`; try "removeConn"

fnordahl replied:
I do see your point, but It feels weird to use "remove" to get a connection to do some work with. But I'll change it, I may try to come up with a alternative name that encompasses both concerns.

bobbyrullo commented:
hey @fnordahl sorry we've neglected this PR so long (if that happens feel free to ping one of us). 

Anyhow, this is mostly good with a few nits. I would like to see some tests though, exercising potential race conditions - for example, what happens when ping checks the liveness of a connection that is no longer in the connection map (since you hold the lock for the map only briefly and then check)

fnordahl commented:
Thank you for your review/comments @bobbyrullo . No worries on time passing, I have had a lot of stuff going on elsewhere latelly so there has unfortunatelly not been much time for me to focus on this. I'm back though and will se this through.

The rationale behind the ping function is that it at first should get a snapshot of the map quickly and not hold the lock for too long. After it has a snapshot of the map it starts to iterate over it using the getConn function. The getConn function does hold a lock, checks the existence of the specific connection requested, removes it from the map if it exists and releases the lock. If it does not exist that fact is returned to the caller. Thus there really should not be any room for races here.

Some scale tests and/or tests of behaviour with actual dead/timed out connections is in place. I'll have a look at how this can be implemented.

fnordahl commented:
Updated version:
- [x] nit: TLSConfig
- [x] *LDAPPool.Ping() -> *LDAPPool.CheckConnections()
- [x] Spelling: allready -> already
- [x] Get() -> Acquire(), getConn(...) -> removeConn(...)

Todo:
- [ ] Make Sync use stop chan

I currently have issues testing and I suspect that something has broken the LDAP connector. I will investigate more tomorrow.

fnordahl commented:
@ericchiang @bobbyrullo I have looked over the Sync stuff in go-oidc, and for the simple tasks the LDAP connector has to perform, it seems like a lot of code to accomplish little. Could I jump straight to attempt at an implementation outlined in #363 ?

Or are there a lot of infrastructure currently missing in Dex for this to be feasible / usable at this time?

philips commented on connector/connector_ldap.go:
```diff
@@ -210,18 +231,131 @@ func (c *LDAPConnector) Register(mux *http.ServeMux, errorURL url.URL) {
+	p.m.Lock()
+	for conn := range p.conns {
+		delete(p.conns, conn)
+		p.m.Unlock()
```
please use defer for unlocks unless there is a strong reason to avoid it

fnordahl replied:
The reasoning behind manual Lock/Unlock in this function is that the call to p.ldapConnect() could make us block while holding the lock. An alternative could be to return nil and let it be up to the user of the function to get a connection himself and later on put it into the pool. At the time of writing it somehow felt better to have all connection related things happen through the pool interface, but there is nothing wrong with doing this separation if you think that fits better.

fnordahl replied:
Could of course also do the actual lifting in a private function with Lock() + defer Lock in place, and do a call to that + return p.ldapConnect() in the public function.

philips commented on connector/connector_ldap.go:
```diff
@@ -210,18 +231,131 @@ func (c *LDAPConnector) Register(mux *http.ServeMux, errorURL url.URL) {
+// another routine is using the connection and the caller should discard the pointer.
+func (p *LDAPPool) removeConn(conn *ldap.Conn) bool {
+	p.m.Lock()
+	defer p.m.Unlock()
```
in this case not having the defer feels cleaner

fnordahl replied:
I'm ok with that. Simple and straight forward function where it is easy to foresee the entire call path with the glance of an eye.

fnordahl commented:
> @ericchiang @bobbyrullo I have looked over the Sync stuff in go-oidc, and for the simple tasks the LDAP connector has to perform, it seems like a lot of code to accomplish little. Could I jump straight to attempt at an implementation outlined in #363 ?
> 
> Or are there a lot of infrastructure currently missing in Dex for this to be feasible / usable at this time?

I believe I found a simpler way of solving it while still following the current scheme. New version up. How does this look?

I still have the issues with something breaking the connector, but I'll continue investigate that and get back with a fix. (I probably have no one else than myself not writing good enough tests to blame, I will make amends.)

Would love some feedback on the intermediate work though.

fnordahl commented:
After doing a binary search I have identified that this commit breaks the LDAP connector: 821b242c83bd1770af0886899be163bdc77749cd

I will investigate further and of course use the wonderful experience of having my code being broken by other commits to improve my tests! :-)

fnordahl commented:
Some more detail: It is the "register-maybe" process that breaks. Manually registering before first login attempt still works.

fnordahl commented:
@philips I have addressed the usage/code-style for locks per your comments.

Should I refactor throughout, moving the parts touching global structures to seperate functions so that the call paths either are obvious at the glimpse of an eye or I can make use of defer Unlock without blocking?

If so, I think @ericchiang have suggested this for me before, so sorry for not listening the first time... ;)

fnordahl commented:
@ericchiang what's your take on this? Are the rest of the functions with locks OK as is, or do you suggest I refactor them? Refactoring would be move parts touching global structures to separate functions that can either be simple enough so you can spot the entire call path at the blink of an eye, or that they contain no blocking calls so Lock+defer Unlock could be used.

ericchiang commented:
@fnordahl sorry to take so long following up.

Regarding the actual code, I'm okay with it not being the most "elegant" implementation if it works. However we're going to need a plan to test this logic (particularly with concurrent requests). Would it be okay to move on to that part?

fnordahl commented:
Thank you for the feedback @ericchiang . Sure, I'll brew up some tests.

fnordahl commented:
And there we have functional stress tests for the LDAP Connection Pool. What do you think of it @ericchiang

ericchiang commented on functional/ldap_test.go:
```diff
@@ -123,3 +125,52 @@ func TestConnectorLDAPHealthy(t *testing.T) {
+	numRoutines := ldapPool.MaxIdleConn + 10
+	var wg sync.WaitGroup
+	for i := 0; i < numRoutines; i++ {
+		wg.Add(1)
```
This is kinda a race. Please just do a `wg.Add(numRountines)`

fnordahl replied:
The calls are executed serially by the for loop, so no room for races here. But sure, I agree that calling the function 40 times is less efficient than once :-)

ericchiang replied:
It's racy because theoretically the `wg.Done` could race with the `wg.Add` (since the latter's in a goroutine). It's more correct this way.

fnordahl replied:
Ack

ericchiang commented on functional/ldap_test.go:
```diff
@@ -123,3 +125,52 @@ func TestConnectorLDAPHealthy(t *testing.T) {
+					if err != nil {
+						t.Errorf("Search request failed. Dead/invalid LDAP connection from pool?: %v", err)
+					}
+					ldapPool.Put(ldapConn)
```
probably don't want to Put if there was an error

fnordahl replied:
The Put function should check the eligibility of the connections for re-use before adding them back to the pool. So it would be a good test. But let's do a explicit test of that instead of randomly putting back known broken connections. Thanks for pointing it out.

ericchiang commented on functional/ldap_test.go:
```diff
@@ -123,3 +125,52 @@ func TestConnectorLDAPHealthy(t *testing.T) {
+
+	// Wait for all operations to complete and check status.
+	wg.Wait()
+	alive, killed := ldapPool.CheckConnections()
```
Why do we care how many are alive?

fnordahl replied:
Checking that there are MaxIdleConn connections active after the test gives us information about two things:
1. Did we indeed have concurrent operations during the test.
2. The test runs more than MaxIdleConn connections concurrently, and checking that the pool stays at MaxIdleConn after the run confirms that the code handling the high watermark of the pool works.

ericchiang replied:
That sound good. But shouldn't a connection be able to error out? Isn't that one of the points of a connection pool (to be able to tolerate bad connections)?

fnordahl replied:
Yes indeed. The connections in this test most likely will never error out. I will create a separate test for the handling of a harsh environment with connections going bad.

ericchiang commented on functional/ldap_test.go:
```diff
@@ -123,3 +125,52 @@ func TestConnectorLDAPHealthy(t *testing.T) {
 	}
 }
+
+func TestLDAPPool(t *testing.T) {
```
Might pick a more descriptive name. Like "TestMultipleLDAPConns". You aren't testing the pool, you're testing concurrent access :)

fnordahl replied:
We do both. In my head, the reason for the test is to verify that we do not have lock contention while exercising the various pool functions concurrently. But sure, I'll find a more descriptive name.

ericchiang commented on functional/ldap_test.go:
```diff
@@ -123,3 +125,52 @@ func TestConnectorLDAPHealthy(t *testing.T) {
+	for i := 0; i < numRoutines; i++ {
+		wg.Add(1)
+		go func() {
+			timeout := time.After(5 * time.Second)
```
This can probably be done with a golang.org/x/net/context object.

fnordahl replied:
Thx! I'll check it out.

ericchiang commented on functional/ldap_test.go:
```diff
@@ -123,3 +125,52 @@ func TestConnectorLDAPHealthy(t *testing.T) {
+func TestLDAPPool(t *testing.T) {
+	server := ldapServer(t)
+	ldapPool := &connector.LDAPPool{
+		MaxIdleConn: 30,
```
Would be cool if this was a benchmark, see how tuning this number effects the throughput. (For a follow up PR)

fnordahl replied:
Indeed, that is a good idea. Didn't think of that. Thx!

ericchiang commented on Documentation/connectors-configuration.md:
```diff
@@ -167,6 +167,10 @@ In addition to `id` and `type`, the `ldap` connector takes the following additio
 
+* maxIdleConn: a `integer`. Maximum number of idle LDAP Connections to keep in connection pool. Default: `5`
+
+* poolCheckTimer: a `integer`. Time in Seconds for periodic liveness-check of connections in pool. Default: `7200` (Every two hours)
```
This doesn't need to be configurable. Let's drop it for brevity.

fnordahl replied:
Done.

ericchiang commented:
one last comment, but other than that lgtm

ericchiang commented:
:+1: as always great work
